src/ucp/rndv/proto_rndv.c

```diff
@@ -278,115 +303,134 @@ ucp_proto_rndv_ctrl_init_priv(const ucp_proto_rndv_ctrl_init_params_t *params,
-    const char *proto_name   = ucp_proto_id_field(params->super.super.proto_id,
-                                                  name);
-    const ucp_proto_perf_range_t *parallel_stages[2];
-    UCS_STRING_BUFFER_ONSTACK(agg_buf, 256);
-    size_t range_max_length;
+    const char *variant_name = ucp_proto_id_field(remote_proto->proto_id, name);
+    size_t num_elems       = 0;
+    const ucp_proto_perf_t *perf_elems[3];
+    ucp_proto_perf_t *ctrl_perf, *remote_perf;
+    UCS_STRING_BUFFER_ONSTACK(perf_name_buf, 256);
+    size_t cfg_thresh, cfg_priority;
+    ucs_linear_func_t overhead;
+    ucp_proto_perf_t *perf;
     ucs_status_t status;
 
-    ucs_string_buffer_appendf(&agg_buf, "%s"UCP_PROTO_PERF_NODE_NEW_LINE"%s",
-                              proto_name, variant_name);
-
-    do {
-        range_max_length = ucs_min(remote_range->max_length, max_length);
-        if (range_max_length < min_length) {
-            continue;
-        }
+    ucs_string_buffer_appendf(&perf_name_buf,
+                              "%s"UCP_PROTO_PERF_NODE_NEW_LINE"%s",
+                              params->ctrl_msg_name,
+                              ucp_proto_perf_name(remote_proto->perf));
+    ucs_trace("%s proto variant: cfg_thresh=%zu cfg_priority=%u", variant_name,
+              remote_proto->cfg_thresh, remote_proto->cfg_priority);
 
-        ucs_trace("%s: max %zu remote-op %s" UCP_PROTO_PERF_FUNC_TYPES_FMT,
-                  proto_name, remote_range->max_length, variant_name,
-                  UCP_PROTO_PERF_FUNC_TYPES_ARG(remote_range->perf));
+    /* Set send_overheads to the time to send and receive RTS message */
+    overhead = ucs_linear_func_make(params->super.overhead * 2, 0.0);
+    status   = ucp_proto_rndv_ctrl_perf(&params->super, rpriv->lane,
+                                        params->ctrl_msg_name, overhead,
+                                        &ctrl_perf);
+    if (status != UCS_OK) {
+        return;
+    }
 
-        parallel_stages[0] = ctrl_range;
-        parallel_stages[1] = remote_range;
-        status             = ucp_proto_init_parallel_stages(
-                ucs_string_buffer_cstr(&agg_buf), min_length, range_max_length,
-                params->perf_bias, parallel_stages, 2, proto_caps);
-        if (status != UCS_OK) {
-            return status;
-        }
+    status = ucp_proto_init_add_memreg_time(
+            &params->super, rpriv->md_map, UCP_PROTO_PERF_FACTOR_LOCAL_CPU,
+            "memory registration", params->super.min_length,
+            params->super.max_length, ctrl_perf);
+    if (status != UCS_OK) {
+        goto out_destroy_ctrl_perf;
+    }
 
-        min_length = range_max_length + 1;
-    } while ((remote_range++)->max_length < max_length);
+    status = ucp_proto_perf_remote(remote_proto->perf, &remote_perf);
+    if (status != UCS_OK) {
+        goto out_destroy_ctrl_perf;
+    }
 
-    return UCS_OK;
-}
+    perf_elems[num_elems++]     = ctrl_perf;
```

Comment:
maybe use ucs_array (onstack) for perf_elems?

Reply:
AFAIU it would require introducing separate type for UCS_ARRAY_ALLOC_ONSTACK or defining separate buffer on stack to store data separately there (for UCS_ARRAY_FIXED_INITIALIZER). Do you think that one of these options are better than usual C array or am I missing a way to detect ucs_array on stack?